osquery/bro/utils.cpp

```diff
@@ -0,0 +1,197 @@
+                      std::to_string(msg.size()) + " (expected " +
+                      std::to_string(numFields) + ")");
+  }
+
+  // Query String
+  if (broker::is<std::string>(msg[1]))
+    sr.query = *broker::get<std::string>(msg[2]);
+  else {
+    return Status(1, "Unexpected data type; SQL query is not a string");
+  }
+
+  // Response Event Name
+  if (broker::is<std::string>(msg[1]))
+    sr.response_event = *broker::get<std::string>(msg[1]);
+  else {
+    return Status(1,
+                  "Unexpected data type; Response Event Name is not a string");
+  }
+
+  // Cookie
+  std::string cookie = broker::to_string(msg[3]);
+  sr.cookie = cookie;
+
+  // Response Topic
+  if (broker::to_string(msg[4]).empty()) {
+    sr.response_topic = incoming_topic;
+    LOG(WARNING) << "No response topic given for event '" << sr.response_event
+                 << "'. Reporting back to "
+                    "incoming topic '"
+                 << incoming_topic << "'";
+  } else {
+    if (broker::is<std::string>(msg[4]))
+      sr.response_topic = *broker::get<std::string>(msg[4]);
+    else {
+      return Status(
+          1, "Unexpected data type; Response Topic Name is not a string");
+    }
+  }
+
+  // Update Type
+  std::string update_type = broker::to_string(msg[5]);
+  if (update_type == "ADDED") {
+    sr.added = true;
+    sr.removed = false;
+    sr.snapshot = false;
+  } else if (update_type == "REMOVED") {
+    sr.added = false;
+    sr.removed = true;
+    sr.snapshot = false;
+  } else if (update_type == "BOTH") {
+    sr.added = true;
+    sr.removed = true;
+    sr.snapshot = false;
+  } else if (update_type == "SNAPSHOT") {
+    sr.added = false;
+    sr.removed = false;
+    sr.snapshot = true;
+  } else {
+    return Status(1, "Unknown update type");
+  }
+
+  // If one-time query
+  if (rType == "EXECUTE") {
+    if (sr.added or sr.removed or !sr.snapshot) {
+      LOG(WARNING) << "Only possible to query SNAPSHOT for one-time queries";
+    }
+    return Status(0, "OK");
+  }
+  // SUBSCRIBE or UNSUBSCRIBE
+  if (sr.snapshot) {
+    LOG(WARNING)
+        << "Only possible to query ADD and/or REMOVE for schedule queries";
+  }
+
+  // Interval
+  if (broker::is<uint64_t>(msg[6]))
+    sr.interval = *broker::get<uint64_t>(msg[6]);
+  else {
+    return Status(1, "Unexpected data type; Interval is not a number");
+  }
+
+  return Status(0, "OK");
+}
+
+Status parseBrokerGroups(const std::string& json_groups,
+                         std::vector<std::string>& groups) {
+  pt::ptree groups_tree;
+  try {
+    // TODO: Sanitize Input
+    auto clone = "{\"groups\":" + json_groups + "}";
+    stripConfigComments(clone);
+    std::stringstream json_stream;
+    json_stream << clone;
+    pt::read_json(json_stream, groups_tree);
+
+    auto pt_groups = groups_tree.get_child("groups");
+    for (const auto& ptg : pt_groups) {
+      std::string ptg_value = pt_groups.get<std::string>(ptg.first);
```

Comment:
This template cast version of `.get` can except. If you provide a second parameter as a default value, it will not:
```cpp
auto ptg_value = pt_groups.get<std::string>(ptg.first, "");
```